homeassistant/core.py

```diff
@@ -330,11 +330,50 @@ def _async_stop_handler(self, *args):
         self.exit_code = 0
         self.loop.create_task(self.async_stop())
 
+    @asyncio.coroutine
+    def _notify(self, message):
+        data = {'message': message,
+                'title': 'Restarting',
+                'notification_id': '{}.restart'.format(DOMAIN)}
+        yield from self.services.async_call(
+            'persistent_notification', 'create', data)
+
+    @asyncio.coroutine
+    def _async_check_config_and_restart(self):
+        """Restart Home Assistant if config is valid.
+
+        This method is a coroutine.
+        """
+        proc = yield from asyncio.create_subprocess_exec(
+            sys.argv[0],
+            '--script',
+            'check_config',
+            stdout=asyncio.subprocess.PIPE)
+        # Wait for the subprocess exit
+        result = yield from proc.wait()
+        if result:
+            _LOGGER.error("check_config failed. Not restarting.")
+            content = yield from proc.stdout.read()
+            content = str(content, 'utf-8')
+            # Print content to stdout so that it will have the colored
+            # pretty-printed error.
+            print(content)
+            # Put error cleaned from color codes in the error log so it
+            # will be visible at the UI.
+            _LOGGER.error(re.sub(r'\033\[[^m]*m', '', content))
+            yield from self._notify('Restart failed')
+            return
+        self.exit_code = RESTART_EXIT_CODE
+        # This notification is unlikely to arrive before HA is stopped.
+        yield from self._notify('Config validated')
+        yield from self.async_stop()
+
     @callback
```

Comment:
Cant you simplify this (get id of `_async_check_config_and_restart` function), by changing this `@callback` into a `@asyncio.coroutine`?